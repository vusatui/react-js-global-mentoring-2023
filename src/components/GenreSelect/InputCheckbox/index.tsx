import {StyledContainer, StyledInput, StyledLabel} from "./styled";

interface InputCheckboxProps {
    className?: string
    name?: string
    value: boolean
    onChange: (v: boolean) => void
}

const InputCheckbox = ({
    className,
    value,
    onChange,
}: InputCheckboxProps) => (
    <StyledContainer
        className={className}
    >
        <StyledInput
            type="checkbox"
            value={String(value)}
            onChange={e => onChange(!!e.target.value)}
        />
        <StyledLabel isChecked={value} />
    </StyledContainer>
);

export default InputCheckbox;