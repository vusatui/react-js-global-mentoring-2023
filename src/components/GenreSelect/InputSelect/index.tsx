import {useState} from "react";

import {
    StyledArrowIcon,
    StyledContainer,
    StyledDropDown,
    StyledDropdownItem,
    StyledInputCheckbox,
    StyleSelectInput
} from "./styled";

interface InputSelectProps {
    className?: string
    label?: string
    name: string
    options: string[]
    value: string[]
    onChange: (v: string[]) => void
}

const InputSelect = ({
    className,
    label = "Select Genre",
    name,
    options,
    value,
    onChange,
}: InputSelectProps) => {
    const [isOpened, setIsOpened] = useState(false);

    const handleContainerClick = () => setIsOpened(!isOpened);

    const handleOptionSelect = (option: string) =>
        onChange(value.includes(option) ? value.filter(item => item !== option) : [...value, option]);

    return (
        <StyledContainer
            className={className}
            onClick={handleContainerClick}
        >
            <StyleSelectInput
                name={name}
                value={value}
                multiple={true}
            >
                {options.map(option =>
                    (<option key={option} value={option}>{option}</option>))}
            </StyleSelectInput>
            <span>{label}</span>
            <StyledArrowIcon isRotated={isOpened} />
            {isOpened && <StyledDropDown onClick={e => e.stopPropagation()}>
                {options.map(option =>
                    (<StyledDropdownItem key={option} onClick={() => handleOptionSelect(option)}>
                        <StyledInputCheckbox value={value.includes(option)} onChange={() => null} />
                        {option}
                    </StyledDropdownItem>))}
            </StyledDropDown>}
        </StyledContainer>
    );
};

export default InputSelect;