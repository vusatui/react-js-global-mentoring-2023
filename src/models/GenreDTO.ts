export interface GenreDTO {
    name: string
}