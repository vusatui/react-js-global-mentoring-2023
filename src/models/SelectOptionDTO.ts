export interface SelectOptionDTO<T> {
    id: string
    label: string
    value: T
}