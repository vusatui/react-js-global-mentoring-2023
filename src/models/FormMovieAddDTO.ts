import {GenreDTO} from "./GenreDTO";

export interface FormMovieAddDTO {
    title: string
    movieUrl: string
    genre: GenreDTO[]
    overview: string
    runtime: string
    rating: string
    releaseDate: string
}