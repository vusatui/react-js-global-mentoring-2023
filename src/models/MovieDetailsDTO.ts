import {GenreDTO} from "./GenreDTO";

export interface MovieDetailsDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
    duration: string
    rating: number
    description: string
}