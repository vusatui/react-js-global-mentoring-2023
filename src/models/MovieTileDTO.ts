import {GenreDTO} from "./GenreDTO";

export interface MovieTileDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
}