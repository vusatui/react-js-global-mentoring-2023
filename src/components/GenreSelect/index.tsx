import {CSSProperties, useMemo, useState} from "react";

import {GenreDTO} from "../../models/GenreDTO";

import InputSelect from "./InputSelect";

interface Props {
    style?: CSSProperties
    className?: string
    genres: GenreDTO[]
    onSelect: (v: GenreDTO[]) => void
}

const GenreSelect = ({
    style = undefined,
    className = undefined,
    genres,
    onSelect,
}: Props) => {
    const [value, setValue] = useState<string[]>([]);

    const genresOptions = useMemo(() => genres.map(genre => genre.name), [genres]);

    const onChange = (v: string[]) => {
        setValue(v);
        onSelect(genres.filter(genre => v.includes(genre.name)));
    };

    return (
        <InputSelect
            style={style}
            className={className}
            name="genre"
            options={genresOptions}
            value={value}
            onChange={onChange}
        />
    );
};

export default GenreSelect;