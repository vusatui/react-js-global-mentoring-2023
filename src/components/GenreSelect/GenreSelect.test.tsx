import {act, render, screen} from "@testing-library/react";
import GenreSelect from "../GenreSelect";
import {GenreDTO} from "../../models/GenreDTO";
import userEvent from "@testing-library/user-event";

const FIRST_ELEMENT_INDEX = 0;

const genresOptions: GenreDTO[] = [
    { name: "Fantastic" },
    { name: "Anime" },
];

const mockHandleSelect = jest.fn();

const Component = () => <GenreSelect genres={genresOptions} onSelect={mockHandleSelect} />;

describe("GenreSelect", () => {

    it("Should renders correctly", () => {
        const { asFragment } = render(<Component />);
        expect(asFragment()).toMatchSnapshot();
        act(() => userEvent.click(screen.getByText("Select Genre")))
        expect(asFragment()).toMatchSnapshot();
    });

    it("Should test that component renders all genres passed in props", () => {
        render(<Component />);
        act(() => userEvent.click(screen.getByText("Select Genre")))
        genresOptions.forEach(
            async ({ name }) => expect(await screen.findAllByText(name))
                .toBeInTheDocument(),
        );
    });

    it("Should test that component highlights a selected genre passed in props", async () => {
        const { asFragment } = render(<Component />);
        act(() => userEvent.click(screen.getByText("Select Genre")))
        const selectItem: HTMLElement = await screen.getAllByRole("option")[FIRST_ELEMENT_INDEX];
        act(() => userEvent.click(selectItem));
        const checkbox: HTMLInputElement | null = selectItem.querySelector("input[type='checkbox']");
        expect(checkbox).toBeInTheDocument();
        expect(checkbox?.value).toBeTruthy();
        expect(asFragment()).toMatchSnapshot();
    });

    it("Test 'onSelect' passes correct genre in arguments", async () => {
        render(<Component />);
        act(() => userEvent.click(screen.getByText("Select Genre")));
        const selectItem: HTMLElement = await screen.getAllByRole("option")[FIRST_ELEMENT_INDEX];
        act(() => userEvent.click(selectItem));
        expect(mockHandleSelect).toBeCalledWith([{ name: "Fantastic" }]);
    });

});