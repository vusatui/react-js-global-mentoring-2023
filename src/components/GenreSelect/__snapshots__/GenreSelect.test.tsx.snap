// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelect Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-hLseeU fYEuDo"
  >
    <select
      class="sc-eDDNvR hFpGcH"
      multiple=""
      name="genre"
    >
      <option
        value="Fantastic"
      >
        Fantastic
      </option>
      <option
        value="Anime"
      >
        Anime
      </option>
    </select>
    <span>
      Select Genre
    </span>
    <svg
      class="sc-jTrPJq dTvPkU"
      fill="none"
      height="12"
      viewBox="0 0 18 12"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M1.13423 0C0.286533 0 -0.176668 0.988581 0.365828 1.63996L8.22615 11.0779C8.6258 11.5578 9.36265 11.558 9.76259 11.0784L17.6322 1.64041C18.1752 0.989135 17.7121 0 16.8642 0H1.13423Z"
        fill="#F65261"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`GenreSelect Should renders correctly 2`] = `
<DocumentFragment>
  <div
    class="sc-hLseeU fYEuDo"
  >
    <select
      class="sc-eDDNvR hFpGcH"
      multiple=""
      name="genre"
    >
      <option
        value="Fantastic"
      >
        Fantastic
      </option>
      <option
        value="Anime"
      >
        Anime
      </option>
    </select>
    <span>
      Select Genre
    </span>
    <svg
      class="sc-jTrPJq klYxgK"
      fill="none"
      height="12"
      viewBox="0 0 18 12"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M1.13423 0C0.286533 0 -0.176668 0.988581 0.365828 1.63996L8.22615 11.0779C8.6258 11.5578 9.36265 11.558 9.76259 11.0784L17.6322 1.64041C18.1752 0.989135 17.7121 0 16.8642 0H1.13423Z"
        fill="#F65261"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="sc-gLDzan ccHUSu"
    >
      <div
        class="sc-iAEyYk fOZodR"
        role="option"
      >
        <div
          class="sc-beqWaB sc-fsQiph ccifld"
        >
          <input
            class="sc-gueYoa jzOEHu"
            type="checkbox"
            value="false"
          />
          <label
            class="sc-dmqHEX jnScJi"
          />
        </div>
        Fantastic
      </div>
      <div
        class="sc-iAEyYk fOZodR"
        role="option"
      >
        <div
          class="sc-beqWaB sc-fsQiph ccifld"
        >
          <input
            class="sc-gueYoa jzOEHu"
            type="checkbox"
            value="false"
          />
          <label
            class="sc-dmqHEX jnScJi"
          />
        </div>
        Anime
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GenreSelect Should test that component highlights a selected genre passed in props 1`] = `
<DocumentFragment>
  <div
    class="sc-hLseeU fYEuDo"
  >
    <select
      class="sc-eDDNvR hFpGcH"
      multiple=""
      name="genre"
    >
      <option
        value="Fantastic"
      >
        Fantastic
      </option>
      <option
        value="Anime"
      >
        Anime
      </option>
    </select>
    <span>
      Select Genre
    </span>
    <svg
      class="sc-jTrPJq klYxgK"
      fill="none"
      height="12"
      viewBox="0 0 18 12"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M1.13423 0C0.286533 0 -0.176668 0.988581 0.365828 1.63996L8.22615 11.0779C8.6258 11.5578 9.36265 11.558 9.76259 11.0784L17.6322 1.64041C18.1752 0.989135 17.7121 0 16.8642 0H1.13423Z"
        fill="#F65261"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="sc-gLDzan ccHUSu"
    >
      <div
        class="sc-iAEyYk fOZodR"
        role="option"
      >
        <div
          class="sc-beqWaB sc-fsQiph ccifld"
        >
          <input
            class="sc-gueYoa jzOEHu"
            type="checkbox"
            value="true"
          />
          <label
            class="sc-dmqHEX gVBKkj"
          />
        </div>
        Fantastic
      </div>
      <div
        class="sc-iAEyYk fOZodR"
        role="option"
      >
        <div
          class="sc-beqWaB sc-fsQiph ccifld"
        >
          <input
            class="sc-gueYoa jzOEHu"
            type="checkbox"
            value="false"
          />
          <label
            class="sc-dmqHEX jnScJi"
          />
        </div>
        Anime
      </div>
    </div>
  </div>
</DocumentFragment>
`;
